src/providers/route.js

```diff
@@ -0,0 +1,244 @@
+        urlWriters[writerName] = function(data) {
+          if (data == null) { data = {}; }
+          let locals = {UrlData: data};
+          return $injector.invoke(writer, {}, locals);
+        }
+      );
+
+      let flashStates = [];
+
+      let service = {
+        readyDeferred: $q.defer(),
+
+        match(urlToMatch) {
+          for (let url of Array.from(urls)) {
+            var match;
+            if ((match = url.compiledUrl.regex.exec(urlToMatch)) !== null) {
+              return {url, regexMatch: match};
+            }
+          }
+          return null;
+        },
+
+        extractData(match, searchData) {
+          if (searchData == null) { searchData = undefined; }
+          let defaults = this.extractDefaultData(match);
+          let path = this.extractPathData(match);
+          searchData = this.extractSearchData(searchData);
+          return ObjectHelper.default(searchData, path, defaults);
+        },
+
+        extractSearchData(searchData) {
+          if (searchData == null) { searchData = $location.search(); }
+          let data = _.clone(searchData);
+          let newData = {};
+
+          for (let key in data) {
+            var value = data[key];
+            let targetKey = _.findKey(tokens, { searchAlias: key });
+            if (targetKey == null) { targetKey = key; }
+
+            if (!tokens[targetKey] || __guard__(types[tokens[targetKey] != null ? tokens[targetKey].type : undefined], x => x.regex.test(value))) {
+
+              if (__guard__(types[tokens[targetKey] != null ? tokens[targetKey].type : undefined], x1 => x1.parser)) {
+                value = $injector.invoke(types[tokens[targetKey].type] != null ? types[tokens[targetKey].type].parser : undefined, null, {token: value});
+              }
+
+              let dataKey = (tokens[targetKey] != null ? tokens[targetKey].statePath : undefined) || targetKey;
+
+              ObjectHelper.set(newData, dataKey, value);
+            }
+          }
+
+          return newData;
+        },
+
+        extractDefaultData(match) {
+          let data = {};
+
+          for (let key in match.url.state) {
+            let value = match.url.state[key];
+            ObjectHelper.set(data, key, (typeof value === 'object' ? _.cloneDeep(value) : value));
+          }
+
+          return data;
+        },
+
+        extractPathData(match) {
+          let data = {};
+          let pathTokens = match.url.compiledUrl.tokens;
+
+          if (pathTokens.length === 0) { return {}; }
+
+          for (let n = 0, end = pathTokens.length-1, asc = 0 <= end; asc ? n <= end : n >= end; asc ? n++ : n--) {
+            let token = match.url.compiledUrl.tokens[n];
+            let value = match.regexMatch[n+1];
+
+            if (types[token.type].parser) { value = $injector.invoke(types[token.type].parser, null, {token: value}); }
+
+            ObjectHelper.set(data, (token.statePath || token.name), value);
+          }
+
+          return data;
+        },
+
```

Comment:
Please change these single line methods to be spread out over multiple lines, like getPersistantStates is, vs invokeUrlWriter